ps.go broken: bug between walk at /proc  and reading stats of process

Sometimes when process disappears at the middle of operations in walking and reading stats, we got that weird error:

```
2016/03/09 19:53:37 /proc/7020: lstat /proc/7020: no such file or directory
--- FAIL: TestPsExecution (0.01s)
    ps_test.go:18: Loading Table fails on some point; lstat /proc/7020: no such file or directory
FAIL
FAIL    github.com/u-root/u-root/cmds/ps    0.008s
```

The most probable functions whose may be the guilty of this are:
- ProcessTable.[readStats](https://github.com/u-root/u-root/blob/master/cmds/ps/ps_linux.go#L94)
- ProcessTable.[LoadTable](https://github.com/u-root/u-root/blob/master/cmds/ps/ps_linux.go#L220)

--- a/cmds/ps/ps.go
+++ b/cmds/ps/ps.go
@@ -193,6 +193,11 @@ func ps(pT ProcessTable) error {
 	for index, p := range pT.table {
 		uid, err := p.GetUid()
 		if err != nil {
+			// It is extremely common for a directory to disappear from
+			// /proc when a process terminates, so ignore those errors.
+			if os.IsNotExist(err) {
+				continue
+			}
 			return err
 		}
 

--- a/cmds/ps/ps.go
+++ b/cmds/ps/ps.go
@@ -57,7 +57,7 @@ func init() {
 // main process table of ps
 // used to make more flexible
 type ProcessTable struct {
-	table    []Process
+	table    []*Process
 	headers  []string // each column to print
 	fields   []string // which fields of process to print, on order
 	fstring  []string // formated strings
@@ -83,7 +83,7 @@ func (pT ProcessTable) Swap(i, j int) {
 
 // Gived a pid, search for a process
 // Returns nil if not found
-func (pT ProcessTable) GetProcess(pid int) (found Process) {
+func (pT ProcessTable) GetProcess(pid int) (found *Process) {
 	for _, p := range pT.table {
 		if p.Pid == strconv.Itoa(pid) {
 			found = p

--- a/cmds/ps/ps_linux.go
+++ b/cmds/ps/ps_linux.go
@@ -218,8 +218,9 @@ func (p process) getTime() string {
 // Walk from the proc files
 // and parsing them
 func (pT *ProcessTable) LoadTable() error {
+	var parsingError error
 	pf := regexp.MustCompile(allProc)
-	err := filepath.Walk(proc, func(name string, fi os.FileInfo, err error) error {
+	filepath.Walk(proc, func(name string, fi os.FileInfo, err error) error {
 		if err != nil {
 			log.Printf("%v: %v\n", name, err)
 			return err
@@ -231,18 +232,13 @@ func (pT *ProcessTable) LoadTable() error {
 		if pf.Match([]byte(fi.Name())) {
 			p := &Process{}
 			if err := p.Parse(fi.Name()); err != nil {
-				log.Print(err)
-				return err
+				parsingError = err
 			}
-			pT.table = append(pT.table, *p)
+			pT.table = append(pT.table, p)
 		}
 
 		return filepath.SkipDir
 	})
 
-	if err == filepath.SkipDir {
-		return nil
-	}
-
-	return err
+	return parsingError
 }

--- a/cmds/ps/ps_linux.go
+++ b/cmds/ps/ps_linux.go
@@ -7,18 +7,15 @@ package main
 import (
 	"fmt"
 	"io/ioutil"
-	"log"
 	"os"
 	"path"
 	"path/filepath"
 	"reflect"
-	"regexp"
 	"strconv"
 	"strings"
 )
 
 const (
-	allProc = "^[0-9]+$"
 	proc    = "/proc"
 	USER_HZ = 100
 )
@@ -91,14 +88,9 @@ type process struct {
 
 // Parse all content of stat to a Process Struct
 // by gived the pid (linux)
-func (p *process) readStat(pid string) error {
-	b, err := ioutil.ReadFile(path.Join(proc, pid, "stat"))
-
-	// if process disappears on the middle of operation
-	// ignore them
-	if os.IsNotExist(err) {
-		return nil
-	} else if err != nil {
+func (p *process) readStat(pid int) error {
+	b, err := ioutil.ReadFile(path.Join(proc, fmt.Sprint(pid), "stat"))
+	if err != nil {
 		return err
 	}
 
@@ -133,13 +125,8 @@ func (p *process) readStat(pid string) error {
 
 // Fetch data from Operating System about process
 // on Linux read data from stat
-func (p *Process) Parse(pid string) error {
-	if err := p.process.readStat(pid); err != nil {
-		return err
-	}
-
-	return nil
-
+func (p *Process) Parse(pid int) error {
+	return p.process.readStat(pid)
 }
 
 // ctty returns the ctty or "?" if none can be found.
@@ -215,30 +202,33 @@ func (p process) getTime() string {
 	return fmt.Sprintf("%02d:%02d:%02d", hrs, mins, secs)
 }
 
-// Walk from the proc files
-// and parsing them
+// Create a ProcessTable containing stats on all processes.
 func (pT *ProcessTable) LoadTable() error {
-	var parsingError error
-	pf := regexp.MustCompile(allProc)
-	filepath.Walk(proc, func(name string, fi os.FileInfo, err error) error {
+	// Match all files and directories directly inside of /proc.
+	matches, err := filepath.Glob(filepath.Join(proc, "*"))
+	if err != nil {
+		return err
+	}
+
+	for _, m := range matches {
+		// Filter out files and directories which are not numbers.
+		pid, err := strconv.Atoi(filepath.Base(m))
 		if err != nil {
-			log.Printf("%v: %v\n", name, err)
+			continue
+		}
+
+		// Parse the process's stat file.
+		p := &Process{}
+		if err := p.Parse(pid); err != nil {
+			// It is extremely common for a directory to disappear from
+			// /proc when a process terminates, so ignore those errors.
+			if os.IsNotExist(err) {
+				continue
+			}
 			return err
 		}
-		if name == proc {
-			return nil
-		}
+		pT.table = append(pT.table, p)
+	}
 
-		if pf.Match([]byte(fi.Name())) {
-			p := &Process{}
-			if err := p.Parse(fi.Name()); err != nil {
-				parsingError = err
-			}
-			pT.table = append(pT.table, p)
-		}
-
-		return filepath.SkipDir
-	})
-
-	return parsingError
+	return nil
 }

--- a/cmds/ps/ps_test.go
+++ b/cmds/ps/ps_test.go
@@ -13,7 +13,6 @@ import (
 // Simple Test trying execute the ps
 // If no errors returns, it's okay
 func TestPsExecution(t *testing.T) {
-	t.Skip("ps is abuggy and this test is breaking travis; skipping")
 	pT := ProcessTable{}
 	if err := pT.LoadTable(); err != nil {
 		t.Fatalf("Loading Table fails on some point; %v", err)

